src/storage-client/src/storage_collections.rs

```diff
@@ -0,0 +1,2816 @@
+            }
+
+            let mut new_read_capability = collection
+                .read_policy
+                .frontier(collection.write_frontier.borrow());
+
+            if id.is_user() {
+                trace!(
+                    %id,
+                    implied_capability = ?collection.implied_capability,
+                    policy = ?collection.read_policy,
+                    write_frontier = ?collection.write_frontier,
+                    ?new_read_capability,
+                    "update_write_frontiers");
+            }
+
+            if PartialOrder::less_equal(&collection.implied_capability, &new_read_capability) {
+                let mut update = ChangeBatch::new();
+                update.extend(new_read_capability.iter().map(|time| (time.clone(), 1)));
+                std::mem::swap(&mut collection.implied_capability, &mut new_read_capability);
+                update.extend(new_read_capability.iter().map(|time| (time.clone(), -1)));
+
+                if !update.is_empty() {
+                    read_capability_changes.insert(*id, update);
+                }
+            }
+        }
+
+        if !read_capability_changes.is_empty() {
+            StorageCollectionsImpl::update_read_capabilities_inner(
+                &self.cmds_tx,
+                &mut self_collections,
+                &mut read_capability_changes,
+            );
+        }
+    }
+
+    async fn downgrade_sinces(&mut self, cmds: Vec<(GlobalId, Antichain<T>)>) {
+        for (id, new_since) in cmds {
+            let since_handle = if let Some(c) = self.since_handles.get_mut(&id) {
+                c
+            } else {
+                // This can happen when someone concurrently drops a collection.
+                trace!("downgrade_sinces: reference to absent collection {id}");
+                continue;
+            };
+
+            if id.is_user() {
+                trace!("downgrading since of {} to {:?}", id, new_since);
+            }
+
+            let epoch = since_handle.opaque().clone();
+            let result = if new_since.is_empty() {
+                // A shard's since reaching the empty frontier is a prereq for
+                // being able to finalize a shard, so the final downgrade should
+                // never be rate-limited.
+                let res = Some(
+                    since_handle
+                        .compare_and_downgrade_since(&epoch, (&epoch, &new_since))
+                        .await,
+                );
+
+                info!(%id, "removing persist handles because the since advanced to []!");
+
+                let _since_handle = self.since_handles.remove(&id).expect("known to exist");
+                let dropped_shard_id = if let Some(shard_id) = self.shard_by_id.remove(&id) {
+                    shard_id
+                } else {
+                    panic!("missing GlobalId -> ShardId mapping for id {id}");
+                };
+                self.txns_shards.remove(&id);
```

Comment:
Currently, `StorageController` does that, because it is responsible for the write side of tables. In the future, we should have a `TableWriter` (or `TableWrites`) that's responsible for this, and take it away from `StorageController`.

Reply:
might be worth a clarifying comment!

Reply:
👌